File: router.go
package velo

import (
	"fmt"
	"strings"
)

const nMethods = 7

func methodIndex(m string) int {
	switch m {
	case "GET":
		return 0
	case "POST":
		return 1
	case "PUT":
		return 2
	case "PATCH":
		return 3
	case "DELETE":
		return 4
	case "HEAD":
		return 5
	case "OPTIONS":
		return 6
	}
	return -1
}

type node struct {
	static map[string]*node
	param  *node
	route  *Route
}

type Router struct {
	exact [nMethods]map[string]*Route
	trees [nMethods]*node
}

func NewRouter() *Router {
	return &Router{}
}

func (r *Router) Add(rt *Route) error {
	mi := methodIndex(rt.Method)
	if mi < 0 {
		return fmt.Errorf("unsupported method %q", rt.Method)
	}
	path := normalize(rt.Pattern)
	if !strings.Contains(path, ":") {
		if r.exact[mi] == nil {
			r.exact[mi] = map[string]*Route{}
		}
		if _, dup := r.exact[mi][path]; dup {
			return fmt.Errorf("duplicate route %s %s", rt.Method, rt.Pattern)
		}
		r.exact[mi][path] = rt
		return nil
	}
	if r.trees[mi] == nil {
		r.trees[mi] = &node{}
	}
	cur := r.trees[mi]
	for _, seg := range segments(path) {
		if strings.HasPrefix(seg, ":") {
			if cur.param == nil {
				cur.param = &node{}
			}
			cur = cur.param
			continue
		}
		if cur.static == nil {
			cur.static = map[string]*node{}
		}
		nxt, ok := cur.static[seg]
		if !ok {
			nxt = &node{}
			cur.static[seg] = nxt
		}
		cur = nxt
	}
	if cur.route != nil {
		return fmt.Errorf("duplicate route %s %s", rt.Method, rt.Pattern)
	}
	cur.route = rt
	return nil
}

func (r *Router) Lookup(method, path string, c *Ctx) *Route {
	mi := methodIndex(method)
	if mi < 0 {
		return nil
	}
	return r.lookupIndex(mi, path, c)
}

func (r *Router) Allows(path string) bool {
	for i := 0; i < nMethods; i++ {
		if r.lookupIndex(i, path, nil) != nil {
			return true
		}
	}
	return false
}

func (r *Router) lookupIndex(mi int, path string, c *Ctx) *Route {
	p := normalize(path)
	if m := r.exact[mi]; m != nil {
		if rt, ok := m[p]; ok {
			if c != nil {
				c.nparams = 0
			}
			return rt
		}
	}
	cur := r.trees[mi]
	if cur == nil {
		return nil
	}
	n := 0
	rest := strings.TrimPrefix(p, "/")
	for len(rest) > 0 {
		var seg string
		if j := strings.IndexByte(rest, '/'); j >= 0 {
			seg, rest = rest[:j], rest[j+1:]
		} else {
			seg, rest = rest, ""
		}
		if nxt, ok := cur.static[seg]; ok {
			cur = nxt
			continue
		}
		if cur.param != nil && seg != "" {
			if n < maxParams && c != nil {
				c.params[n] = seg
			}
			n++
			cur = cur.param
			continue
		}
		return nil
	}
	if cur.route == nil {
		return nil
	}
	if c != nil {
		c.nparams = n
	}
	return cur.route
}

func normalize(p string) string {
	if len(p) > 1 && p[len(p)-1] == '/' {
		return p[:len(p)-1]
	}
	if p == "" {
		return "/"
	}
	return p
}

func segments(p string) []string {
	t := strings.Trim(p, "/")
	if t == "" {
		return nil
	}
	return strings.Split(t, "/")
}

File: src/router.rs
use crate::parser::{Ctx, Method, Route, MAX_PARAMS, N_METHODS};
use std::collections::HashMap;
use std::hash::{BuildHasherDefault, Hasher};

#[derive(Default)]
pub struct Fnv(u64);

impl Hasher for Fnv {
    fn finish(&self) -> u64 {
        self.0
    }
    fn write(&mut self, bytes: &[u8]) {
        let mut h = if self.0 == 0 { 0xcbf29ce484222325 } else { self.0 };
        for b in bytes {
            h ^= *b as u64;
            h = h.wrapping_mul(0x100000001b3);
        }
        self.0 = h;
    }
}

type FnvMap<K, V> = HashMap<K, V, BuildHasherDefault<Fnv>>;

#[derive(Default)]
struct Node {
    static_kids: Vec<(String, usize)>,
    param: Option<usize>,
    route: Option<usize>,
}

pub struct Router {
    exact: Vec<FnvMap<String, usize>>,
    roots: Vec<Option<usize>>,
    nodes: Vec<Node>,
}

impl Default for Router {
    fn default() -> Self {
        Self::new()
    }
}

impl Router {
    pub fn new() -> Router {
        Router {
            exact: (0..N_METHODS).map(|_| FnvMap::default()).collect(),
            roots: vec![None; N_METHODS],
            nodes: Vec::new(),
        }
    }

    pub fn build(routes: &[Route]) -> Result<Router, String> {
        let mut r = Router::new();
        for (i, rt) in routes.iter().enumerate() {
            r.add(rt, i)?;
        }
        Ok(r)
    }

    fn node(&mut self) -> usize {
        self.nodes.push(Node::default());
        self.nodes.len() - 1
    }

    fn add(&mut self, rt: &Route, idx: usize) -> Result<(), String> {
        let mi = rt.method.index();
        let path = normalize(&rt.pattern).to_string();
        let dup = || format!("duplicate route {} {}", rt.method.name(), rt.pattern);
        if !path.contains(':') {
            if self.exact[mi].insert(path, idx).is_some() {
                return Err(dup());
            }
            return Ok(());
        }
        let mut cur = match self.roots[mi] {
            Some(n) => n,
            None => {
                let n = self.node();
                self.roots[mi] = Some(n);
                n
            }
        };
        for seg in path.trim_matches('/').split('/') {
            if seg.starts_with(':') {
                cur = match self.nodes[cur].param {
                    Some(n) => n,
                    None => {
                        let n = self.node();
                        self.nodes[cur].param = Some(n);
                        n
                    }
                };
                continue;
            }
            cur = match self.nodes[cur].static_kids.iter().find(|(s, _)| s == seg) {
                Some(&(_, n)) => n,
                None => {
                    let n = self.node();
                    self.nodes[cur].static_kids.push((seg.to_string(), n));
                    n
                }
            };
        }
        if self.nodes[cur].route.is_some() {
            return Err(dup());
        }
        self.nodes[cur].route = Some(idx);
        Ok(())
    }

    pub fn lookup<'a>(&self, m: Method, path: &'a str, c: &mut Ctx<'a>) -> Option<usize> {
        self.lookup_index(m.index(), path, Some(c))
    }

    pub fn allows(&self, path: &str) -> bool {
        (0..N_METHODS).any(|i| self.lookup_index(i, path, None).is_some())
    }

    fn lookup_index<'a>(
        &self,
        mi: usize,
        path: &'a str,
        mut c: Option<&mut Ctx<'a>>,
    ) -> Option<usize> {
        let p = normalize(path);
        if let Some(&idx) = self.exact[mi].get(p) {
            if let Some(c) = c.as_deref_mut() {
                c.nparams = 0;
            }
            return Some(idx);
        }
        let mut cur = self.roots[mi]?;
        let mut n = 0;
        let mut rest = p.strip_prefix('/').unwrap_or(p);
        while !rest.is_empty() {
            let (seg, next) = match rest.find('/') {
                Some(j) => (&rest[..j], &rest[j + 1..]),
                None => (rest, ""),
            };
            rest = next;
            if let Some(&(_, nx)) = self.nodes[cur].static_kids.iter().find(|(s, _)| s == seg) {
                cur = nx;
                continue;
            }
            match self.nodes[cur].param {
                Some(nx) if !seg.is_empty() => {
                    if n < MAX_PARAMS {
                        if let Some(c) = c.as_deref_mut() {
                            c.params[n] = seg;
                        }
                    }
                    n += 1;
                    cur = nx;
                }
                _ => return None,
            }
        }
        let idx = self.nodes[cur].route?;
        if let Some(c) = c {
            c.nparams = n.min(MAX_PARAMS);
        }
        Some(idx)
    }
}

fn normalize(p: &str) -> &str {
    if p.is_empty() {
        return "/";
    }
    if p.len() > 1 && p.ends_with('/') {
        return &p[..p.len() - 1];
    }
    p
}
